crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+                            "no current matching server cert found");
+    return 1;
+}
+
+/*
+ * Exceptional handling for 3GPP TS 33.310, only to use for IP and if the ctx
+ * option is explicitly set: use self-signed certificates from extraCerts as
+ * trust anchor to validate server cert - provided it also can validate the
+ * newly enrolled certificate
+ */
+static int srv_cert_valid_3gpp(OSSL_CMP_CTX *ctx, X509 *scrt,
+                               const OSSL_CMP_MSG *msg)
+{
+    int valid = 0;
+    X509_STORE *store = X509_STORE_new();
+
+    if (store != NULL /* store does not include CRLs */
+            && ossl_cmp_X509_STORE_add1_certs(store, msg->extraCerts,
+                                              1 /* self-signed only */))
+        valid = OSSL_CMP_validate_cert_path(ctx, store, scrt);
+
+    if (valid) {
+        /*
+         * verify that the newly enrolled certificate (which is assumed to have
+         * rid == 0) can also be validated with the same trusted store
+         */
+        EVP_PKEY *privkey = OSSL_CMP_CTX_get0_newPkey(ctx, 1);
+        OSSL_CMP_CERTRESPONSE *crep =
+            ossl_cmp_certrepmessage_get0_certresponse(msg->body->value.ip, 0);
+        X509 *newcrt = ossl_cmp_certresponse_get1_certificate(privkey, crep);
+        /*
+         * maybe better use get_cert_status() from cmp_client.c, which catches
+         * errors
+         */
+        valid = OSSL_CMP_validate_cert_path(ctx, store, newcrt);
+        X509_free(newcrt);
+    }
+    X509_STORE_free(store);
+    return valid;
+}
+
+/* Find and validate any potentially usable server cert */
+static X509 *find_srvcert(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+{
+    X509 *scrt = NULL;
+    int valid = 0;
+    GENERAL_NAME *sender = msg->header->sender;
+
+    if (sender == NULL || msg->body == NULL)
+        return 0; /* other NULL cases already have been checked */
+    if (sender->type != GEN_DIRNAME) {
+        CMPerr(0, CMP_R_SENDER_GENERALNAME_TYPE_NOT_SUPPORTED);
+        return NULL; /* FR#42: support for more than X509_NAME */
+    }
+
+    /*
+     * valid scrt, matching sender name, found earlier in transaction, will be
+     * used for validating any further msgs where extraCerts may be left out
+     */
+    (void)ERR_set_mark();
+    if (ctx->validatedSrvCert != NULL
+            && cert_acceptable(ctx->validatedSrvCert, msg, ctx->trusted)) {
+        (void)ERR_pop_to_mark();
+        scrt = ctx->validatedSrvCert;
+        valid = 1;
+    } else {
+        STACK_OF(X509) *found_crts;
+        char *sname;
+        int i;
+
+        (void)ERR_pop_to_mark();
+
+        /* sk_TYPE_find to use compfunc X509_cmp, not ptr comparison */
+        if ((found_crts = sk_X509_new_null()) == NULL)
+            return NULL;
+
+        /* tentatively set error, which allows accumulating diagnostic info */
+        (void)ERR_set_mark();
+        CMPerr(0, CMP_R_NO_VALID_SERVER_CERT_FOUND);
```

Comment:
This looks very strange. AFAICT we are setting an error unconditionally, and then later removing it if we decide we were successful. Doesn't quite seem the right way around!

Reply:
The called functions will fill the error stack if failing. For such case we want to put some context and explanations in front.

Reply:
This doesn't seem right to me. We don't do this *anywhere* else in OpenSSL AFAIK. Everywhere else errors are treated like a stack with the most specific added first, and the most generic added last. It seems very confusing to do this the other way around.

Reply:
I disagree with @mattcaswell that having the most generic error first is confusing, generally speaking. However, having *one* subsystem doing things in the reverse order compared to everything else, *that* is confusing!

In the end, it comes down to display rather than insertion. `ERR_print_errors()` prints things in insertion order, but an application can choose to display things in a different order with creative use of `ERR_print_errors_cb()` (which totally destroys what you're attempting here).

Reply:
> I disagree with @mattcaswell that having the most generic error first is confusing, generally speaking. However, having one subsystem doing things in the reverse order compared to everything else, that is confusing!

I think that was really what I was trying to say. Only you said it better.

Reply:
See also the more general issue being discussed here: https://github.com/openssl/openssl/issues/10537

Reply:
According to the outcome of the discussion 
I've move diagnostic info on search for acceptable cert from to error queue to the CMP log.
Doing so I realized some (partly related) optimizations.